core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -279,6 +277,104 @@ object Uri {
+
+    override def equals(obj: Any): Boolean = obj match {
+      case s: String => doEquals(Uri.Path.fromString(s))
+      case p: Path => doEquals(p)
+      case _ => false
+    }
+
+    private def doEquals(path: Path): Boolean =
+      this.segments == path.segments && path.absolute == this.absolute && path.endsWithSlash == this.endsWithSlash
+
+    override def hashCode(): Int =
+      31 * segments.hashCode() + System.identityHashCode(absolute) + System.identityHashCode(
+        endsWithSlash)
+
+    def render(writer: Writer): writer.type = {
+      val start = if (absolute) "/" else ""
+      writer << start << segments.iterator.mkString("/")
+      if (endsWithSlash) writer << "/" else writer
+    }
+
+    override val renderString: String = super.renderString
+
+    def /(segment: String): Path = addSegment(segment)
+    def addSegment(segment: String): Path =
+      addSegments(List(segment))
+    def addSegments(value: List[String]): Path =
+      Path(this.segments ++ value, absolute = absolute || this.segments.isEmpty)
+
+    def normalize: Path = Path(segments.filterNot(_.isEmpty))
+
+    /* Merge paths per RFC 3986 5.2.3 */
+    def merge(path: Path): Path = {
+      val merge = segments.init
+      Path(merge ++ path.segments, absolute = absolute, endsWithSlash = path.endsWithSlash)
+    }
+
+    def startsWith(path: Path): Boolean = segments.startsWith(path.segments)
+
+    def startsWithString(path: String): Boolean = startsWith(Path.fromString(path))
+
+    def indexOf(path: Path): Option[Int] =
+      Some(segments.indexOfSlice(path.segments)).filterNot(_ == -1)
+
+    def indexOfString(path: String): Option[Int] = indexOf(Path.fromString(path))
+
+    def splitAt(idx: Int): (Path, Path) = {
+      val (start, end) = segments.splitAt(idx)
+      Path(start, absolute) -> Path(end, endsWithSlash)
+    }
+  }
+
+  object Path {
+    val empty = Path(Nil)
+    val Root = Path(Nil, absolute = true)
+
+    /**
+      * This constructor allows you to construct the path directly.
+      * Each path segment needs to be encoded for it to be used here.
+      *
+      * @param segments the segments that this path consists of. MUST be Urlencoded.
+      * @param absolute if the path is absolute. I.E starts with a "/"
+      * @param endsWithSlash if the path is a "directory", ends with a "/"
+      * @return a Uri.Path that can be used in Uri, or by itself.
+      */
+    def apply(
+        segments: List[String],
+        absolute: Boolean = false,
+        endsWithSlash: Boolean = false): Path =
+      new Path(segments, absolute, endsWithSlash)
+
+    def unapply(path: Path): Some[(List[String], Boolean, Boolean)] =
+      Some((path.segments, path.absolute, path.endsWithSlash))
+
+    def fromString(fromPath: String): Path = fromPath match {
+      case "" => empty
+      case "/" => Root
+      case pth =>
+        Path(
+          segments = pth
+            .split("/")
+            .filterNot(_.trim.isEmpty)
```

Comment:
What in the model depends on this?  `//` is not semantically equivalent to `/` in a Uri, even if they're often the same.

Reply:
Maybe they are not, but I have _never_ seen a uri with empty segments that was not a user mistake. 
If we cant remove empty segments here then perhaps we need to allow the pattern matching interpreter to allow to drop empty segments.

Reply:
Maybe its enough having the `normalize` method, so users can choose to this themselves.

Reply:
I tried removing it now, and most of the tests broke. So something is dependant on this, more investigation is needed.

Reply:
Ah, i understand. never mind, will fix